candidate-agreement/src/table.rs

```diff
@@ -0,0 +1,986 @@
+			signalled_bad: self.indicated_bad(),
+		}
+	}
+}
+
+// validator metadata
+struct ValidatorData<C: Context> {
+	proposal: Option<(C::Digest, C::Signature)>,
+	known_statements: HashSet<StatementTrace<C::ValidatorId, C::Digest>>,
+}
+
+/// Create a new, empty statement table.
+pub fn create<C: Context>() -> Table<C> {
+	Table {
+		validator_data: HashMap::default(),
+		detected_misbehavior: HashMap::default(),
+		candidate_votes: HashMap::default(),
+	}
+}
+
+/// Stores votes
+#[derive(Default)]
+pub struct Table<C: Context> {
+	validator_data: HashMap<C::ValidatorId, ValidatorData<C>>,
+	detected_misbehavior: HashMap<C::ValidatorId, <C as ResolveMisbehavior>::Misbehavior>,
+	candidate_votes: HashMap<C::Digest, CandidateData<C>>,
+}
+
+impl<C: Context> Table<C> {
+	/// Produce a set of proposed candidates.
+	///
+	/// This will be at most one per group, consisting of the
+	/// best candidate for each group with requisite votes for inclusion.
+	pub fn proposed_candidates(&self, context: &C) -> Vec<C::Candidate> {
+		use std::collections::BTreeMap;
+		use std::collections::btree_map::Entry as BTreeEntry;
+
+		let mut best_candidates = BTreeMap::new();
+		for candidate_data in self.candidate_votes.values() {
+			let group_id = &candidate_data.group_id;
+			let (validity_t, availability_t) = context.requisite_votes(group_id);
+
+			if !candidate_data.can_be_included(validity_t, availability_t) { continue }
+			let candidate = &candidate_data.candidate;
+			match best_candidates.entry(group_id.clone()) {
+				BTreeEntry::Occupied(mut occ) => {
+					let candidate_ref = occ.get_mut();
+					if *candidate_ref < candidate {
+						*candidate_ref = candidate;
+					}
+				}
+				BTreeEntry::Vacant(vacant) => { vacant.insert(candidate); },
+			}
+		}
+
+		best_candidates.values().map(|v| C::Candidate::clone(v)).collect::<Vec<_>>()
+	}
+
+	/// Get an iterator of all candidates with a given group.
+	// TODO: impl iterator
+	pub fn candidates_in_group<'a>(&'a self, group_id: C::GroupId)
+		-> Box<Iterator<Item=&'a CandidateData<C>> + 'a>
+	{
+		Box::new(self.candidate_votes.values().filter(move |c| c.group_id == group_id))
+	}
+
+	/// Drain all misbehavior observed up to this point.
+	pub fn drain_misbehavior(&mut self) -> HashMap<C::ValidatorId, <C as ResolveMisbehavior>::Misbehavior> {
+		::std::mem::replace(&mut self.detected_misbehavior, HashMap::new())
+	}
+
+	/// Import a signed statement. Signatures should be checked for validity, and the
+	/// sender should be checked to actually be a validator.
+	///
+	/// This can note the origin of the statement to indicate that he has
+	/// seen it already.
+	pub fn import_statement(
+		&mut self,
+		context: &C,
+		statement: SignedStatement<C::Candidate, C::Digest, C::ValidatorId, C::Signature>,
+		from: Option<C::ValidatorId>
+	) -> Option<Summary<C::Digest, C::GroupId>> {
+		let SignedStatement { statement, signature, sender: signer } = statement;
+
+		let trace = match statement {
```

Comment:
looks a bit boilerplatey... no way of combining `Statement` and `StatementTrace`?

Reply:
`StatementTrace<V, D>` is pretty close to `Statement<V, (V, D)>` but it's a little weird. I prefer having the separation.